pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -393,94 +395,90 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
+			fns = append(fns, func() (bool, error) {
+				// delete and recreate failed pods
+				if isFailed(replicas[i]) {
+					ssc.recorder.Eventf(set, v1.EventTypeWarning, "RecreatingFailedPod",
+						"StatefulSet %s/%s is recreating failed Pod %s",
+						set.Namespace,
+						set.Name,
+						replicas[i].Name)
+					if err := ssc.podControl.DeleteStatefulPod(set, replicas[i]); err != nil {
+						return true, err
+					}
+					replicaOrd := i + getStartOrdinal(set)
+					replicas[i] = newVersionedStatefulSetPod(
+						currentSet,
+						updateSet,
+						currentRevision.Name,
+						updateRevision.Name,
+						replicaOrd)
 				}
-			}
-			if err := ssc.podControl.CreateStatefulPod(ctx, set, replicas[i]); err != nil {
-				return &status, err
-			}
-			status.Replicas++
-			if getPodRevision(replicas[i]) == currentRevision.Name {
-				status.CurrentReplicas++
-			}
-			if getPodRevision(replicas[i]) == updateRevision.Name {
-				status.UpdatedReplicas++
-			}
-			// if the set does not allow bursting, return immediately
+				// If we find a Pod that has not been created we create the Pod
+				if !isCreated(replicas[i]) {
+					if utilfeature.DefaultFeatureGate.Enabled(features.StatefulSetAutoDeletePVC) {
+						if isStale, err := ssc.podControl.PodClaimIsStale(set, replicas[i]); err != nil {
+							return true, err
+						} else if isStale {
+							// If a pod has a stale PVC, no more work can be done this round.
+							return true, err
+						}
+					}
+					return true, ssc.podControl.CreateStatefulPod(ctx, set, replicas[i])
+				}
+				return true, nil
+			})
+
 			if monotonic {
-				return &status, nil
+				// if the set does not allow bursting, end loop - no point stacking more functions as they won't be executed anyway.
+				break
 			}
-			// pod created, no more work possible for this round
-			continue
 		}
 
 		// If the Pod is in pending state then trigger PVC creation to create missing PVCs
 		if isPending(replicas[i]) {
-			klog.V(4).Infof(
-				"StatefulSet %s/%s is triggering PVC creation for pending Pod %s",
-				set.Namespace,
-				set.Name,
-				replicas[i].Name)
-			if err := ssc.podControl.createMissingPersistentVolumeClaims(ctx, set, replicas[i]); err != nil {
-				return &status, err
-			}
+			fns = append(fns, func() (bool, error) {
+				klog.V(4).Infof(
+					"StatefulSet %s/%s is triggering PVC creation for pending Pod %s",
+					set.Namespace,
+					set.Name,
+					replicas[i].Name)
+				if err := ssc.podControl.createMissingPersistentVolumeClaims(ctx, set, replicas[i]); err != nil {
+					return true, err
+				}
+				return false, nil
+			})
 		}
 		// If we find a Pod that is currently terminating, we must wait until graceful deletion
 		// completes before we continue to make progress.
 		if isTerminating(replicas[i]) && monotonic {
 			logger.V(4).Info("StatefulSet is waiting for Pod to Terminate",
 				"statefulSet", klog.KObj(set), "pod", klog.KObj(replicas[i]))
-			return &status, nil
+			break
```

Comment:
If we break here, we should leave the function right after "for _, fn := range fns" as before this PR.

Reply:
Fixed - added action in this case as well. I think it makes sense for monotonic to preserve the order of logging as well.